c++11/include/opentracing/propagation.h

```diff
@@ -0,0 +1,135 @@
+  // i.e., the new Span is somehow caused by the work of its FollowsFromRef.
+  //
+  // All of the following could be valid timing diagrams for children that
+  // "FollowFrom" a parent.
+  //
+  //     [-Parent Span-]  [-Child Span-]
+  //
+  //
+  //     [-Parent Span--]
+  //      [-Child Span-]
+  //
+  //
+  //     [-Parent Span-]
+  //                 [-Child Span-]
+  //
+  // See http://opentracing.io/spec/
+  //
+  // See opentracing.FollowsFrom()
+  FollowsFromRef = 2
+};
+
+// Carrier format values.
+enum class CarrierFormat {
+  // OpenTracingBinary encodes the SpanContext for propagation as opaque
+  // binary data.
+  OpenTracingBinary = 1,  // RESERVED, NOT IMPLEMENTED
+
+  // HTTPHeaders represents SpanContexts as HTTP header string pairs.
+  //
+  // The HTTPHeaders format requires that the keys and values be valid
+  // as HTTP headers as-is (i.e., character casing may be unstable and
+  // special characters are disallowed in keys, values should be
+  // URL-escaped, etc).
+  //
+  // For Tracer::Inject(): the carrier must be a `TextMapReader`.
+  //
+  // For Tracer::Extract(): the carrier must be a `TextMapWriter`.
+  //
+  // For example, Inject():
+  //
+  //   std::vector<std::pair<std::string, std::string>> *headers = ...;
+  //   if (!span.tracer().Inject(span, CarrierFormat::HTTPHeadersCarrier,
+  // 	  			   make_ordered_string_pairs_writer(headers))) {
+  //     throw error("inject failed");
+  //   }
+  //
+  // Or Extract():
+  //
+  //   SpanContext extracted;
+  //   extracted = Tracer::Global().Extract(CarrierFormat::HTTPHeadersCarrier,
+  //                                        make_ordered_string_pairs_reader(*headers));
+  //   auto span = Tracer::Global().StartSpan("op", { ChildOf(extracted) });
+  //
+  HTTPHeaders = 2,
+
+  // TextMap encodes the SpanContext as key:value pairs.
+  //
+  // The TextMap format is similar to the HTTPHeaderes format,
+  // without restrictions on the character set.
+  //
+  // For Tracer::Inject(): the carrier must be a `TextMapReader`.
+  //
+  // For Tracer::Extract(): the carrier must be a `TextMapWriter`.
+  //
+  // See the HTTPHeaders examples.
+  TextMap = 3
+};
+
+// Base class for implementation-dependent Tracer::Inject carrier-type adapter.
+class CarrierReader {
+ public:
+  virtual ~CarrierReader() = default;
+};
+
+// Basic foundation for OpenTracing basictracer-compatible carrier readers.
+class BasicCarrierReader : public CarrierReader {
+ public:
+  virtual void ForeachKey(
+      std::function<void(const std::string& key, const std::string& value)> f)
+      const = 0;
+};
+
+// Base class for implementation-dependent Tracer::Extract carrier-type adapter.
+class CarrierWriter {
+ public:
+  virtual ~CarrierWriter() = default;
+};
+
+// Basic foundation for OpenTracing basictracer-compatible carrier writers.
+class BasicCarrierWriter : public CarrierWriter {
+ public:
+  virtual void Set(const std::string& key, const std::string& value) const = 0;
```

Comment:
Should `Set` return an error-code that allows a user to bail out of the writer if the key-value pair can't be written? I found the lack of such a return code inconvenient when writing the [CarrierWriter](https://github.com/rnburn/nginx-opentracing/blob/master/module/src/inject_span_context.cpp#L92) for Nginx's instrumentation, since setting a header value can fail and you have to otherwise manage the error separately or use exceptions break out of the function.